[SERF-1828] Send sentry event via hook only of error is provided

laynax opened the pull request:
## Description

Sentry hook is now triggered only if an error exist now.

Extra points:
* new test was created.
* redundant if condition was removed.
* sentry client initialization was replaced with package level init. this is suggested by library owners, makes it possible to unit test and simplify `sentry.go` code.

## Testing considerations

* run `docker-compose run --rm sdk mage test:run`
* Make sure stack trace is visible in sentry dashboard.

## Checklist

<!-- Please make sure all of the items below are checked before requesting a review: -->

- [x] Prefixed the PR title with the JIRA ticket code <!-- e.g. `[JIRXXX] Add <XYZ> repository` -->
- [x] Performed simple, atomic commits with [good commit messages][commit messages]
- [x] Verified that the commit history is linear and commits are squashed as necessary
- [x] Thoroughly tested the changes in `development` and/or `staging`
- [x] Updated the `README.md` as necessary

<!-- Feel free to open a draft PR to get feedback early. -->

## Related links

* [SERF-1828](https://scribdjira.atlassian.net/browse/SERF-1828) Go SDK Logger: Send Sentry event via Hook only of error is provided

Neurostep commented on pkg/tracking/sentry.go:
```diff
@@ -38,20 +38,22 @@ func (hook *Hook) Levels() []logrus.Level {
-			Stacktrace: stacktrace,
-		})
+	entryError, ok := entry.Data[logrus.ErrorKey].(error)
+	if !ok || entryError == nil {
```
[q]: since we are changing the semantics of the function, can we cover this in unit tests?

Neurostep commented on README.md:
```diff
@@ -573,6 +574,9 @@ func main() {
 	}
+
+	err = errors.New("new error")
+	logger.WithError(err).Errorf("trying sentry error functionality")
```
[q]: can we also add `WithError` to all `Fatal` calls?

In a separate commit, we can update all occurrences of `log.Fatal` to include `WithError(err)` in the README., wdyt?

laynax replied:
I don't understand the change request. "add this to all fatal calls"? can you suggest change?

Neurostep replied:
@laynax Sorry, let me elaborate here. In the README.md file for this section, we have a couple of [log.Fatalf calls](https://github.com/scribd/go-sdk/blob/c8fad6a2f362691f16e4bd27e0aec016a0f25f11/README.md?plain=1#L570-L574). `Fatalf` indicates a major failure that, probably, should be tracked by Sentry.

But now I see that those `Fatalf` calls are from the `log` package. Then, you can ignore my question here 😃 

>I don't understand the change request

Nah, that was not a change request, it is just a comment/question 👍

Neurostep commented on README.md:
```diff
@@ -542,14 +542,15 @@ production:
 
 import (
 	"log"
+	"error"
```
Here is probably a typo, the correct package name is `errors`
```suggestion
	"errors"
```

Neurostep commented on pkg/tracking/sentry.go:
```diff
@@ -116,20 +113,19 @@ func NewSentryHook(config *Config) (*Hook, error) {
-	if len(hook.levels) == 0 {
-		hook.levels = logrus.AllLevels
+	levels := []logrus.Level{
+		logrus.InfoLevel,
```
I believe `logrus.InfoLevel` should be removed from the slice, right?

Neurostep commented on pkg/tracking/sentry_test.go:
```diff
@@ -1,35 +1,60 @@
 	hook, err := NewSentryHook(&config)
 	assert.NoError(t, err)
-	assert.Equal(t, []logrus.Level{logrus.PanicLevel, logrus.FatalLevel, logrus.ErrorLevel}, hook.Levels())
+	assert.Equal(t, []logrus.Level{logrus.InfoLevel,
```
I believe `logrus.InfoLevel` should be removed from the slice, right?

Commit:
fix: Fire sentry event fix

sentry event is now only fired in case there is an error.
sentry client was replaced with package initialization as suggested by library owner. this also made it possible to write unit test.
redundant if check was removed.
[SERF-1828] Go SDK Logger: Send Sentry event via Hook only of error is provided

Commit:
test: Add sentry hook test

replaced client with hub to make code testable.
refactor hardcoded line.

[SERF-1828] Go SDK Logger: Send Sentry event via Hook only of error is provided

Commit:
chore: Update readme

updated sentry error tracking documentation
[SERF-1828] Go SDK Logger: Send Sentry event via Hook only of error is provided

Neurostep reviewed (approved):
LGTM :rocket
